crates/wadm-types/src/validation.rs

```diff
@@ -722,6 +723,47 @@ pub fn is_valid_label_name(name: &str) -> bool {
             .all(|c| c.is_ascii_alphanumeric() || c == '-' || c == '_' || c == '.')
 }
 
+/// Checks whether a manifest contains "duplicate" links.
+///
+/// Multiple links from the same source with the same name, namespace and package
+/// are considered duplicate links.
+fn check_duplicate_links(manifest: &Manifest) -> Vec<ValidationFailure> {
+    let mut failures = Vec::new();
+    let mut link_ids = HashSet::new();
+    for component in manifest.components() {
+        for link in component.links() {
+            if let TraitProperty::Link(LinkProperty {
+                name,
+                namespace,
+                package,
+                ..
+            }) = &link.properties
+            {
```

Comment:
```suggestion
    for component in manifest.components() {
        let mut link_ids = HashSet::new();
        for link in component.links() {
            if let TraitProperty::Link(LinkProperty {
                name,
                namespace,
                package,
                ..
            }) = &link.properties
            {
```

nit: if we move the declaration of the `link_ids` here inside of the loop, it would reset on each iteration and you could omit the `component.name.clone()` when inserting into the hashset.